helpers/nextAuth.ts

```diff
@@ -3,14 +3,74 @@ import { Account, User } from 'next-auth'
 import { LoggedRequest } from '../@types/helpers'
 import { Request, Response } from 'express'
 import { NextApiResponse } from 'next'
-import DiscordProvider from 'next-auth/providers/discord'
 import { getUserSession } from './getUserSession'
+import { get } from 'lodash'
+import { login, signup } from '../graphql/resolvers/authController'
 import prisma from '../prisma'
+import DiscordProvider from 'next-auth/providers/discord'
+import CredentialsProvider from 'next-auth/providers/credentials'
+
+type Credentials =
+  | Record<'username' | 'password' | 'email' | 'firstName' | 'lastName', string>
+  | undefined
+
+export const authorize =
+  (req: LoggedRequest & Request, res: NextApiResponse & Response) =>
+  async (credentials: Credentials) => {
+    const context = { req, res }
+
+    const username = get(credentials, 'username')
+    const password = get(credentials, 'password')
+    const email = get(credentials, 'email')
+    const firstName = get(credentials, 'firstName')
+    const lastName = get(credentials, 'lastName')
+
+    const isSignUpFlow = username && email && firstName && lastName && password
+    const isLoginFlow = username && password
+
+    const user = isSignUpFlow
+      ? await signup(
+          undefined,
+          {
+            username,
+            email,
+            // Will be uncommented when signup takes password as argument
+            // password,
+            firstName,
+            lastName
+          },
+          context
+        )
+      : isLoginFlow && (await login(undefined, { username, password }, context))
```

Comment:
nit: for complex logic like this one, it is better for readability if you explicitly used `if/else` instead of ternary operator. Keep Ternary operator to simple and obvious if/else use cases.

Reply:
I chose immutability over readability. We could argue that even if it is mutable, it's not mutated in any other place than the `if/else`.